infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/send/logreader/RoutingQueuesSnapshotReader.java

```diff
@@ -1,29 +1,288 @@
+
+    // Stream tag which will be followed to fetch any writes to the snapshot sync queue for this destination
+    private String streamTagFollowed;
+
+    private ExecutorService executorService = Executors.newSingleThreadExecutor();
+
+    private boolean endMarkerReached = false;
+
+    // Timestamp of the last read entry in the queue.  Every subsequent batch read will continue from this timestamp
+    private long lastReadTimestamp;
+
+    // UUID of the replicated queue on the receiver(Sink)
+    private UUID replicatedQueueId;
+
+    // UUID of the stream which contains snapshot end markers
+    private UUID endMarkerStreamId;
+
     public RoutingQueuesSnapshotReader(CorfuRuntime corfuRuntime, LogReplicationSession session,
                                        LogReplicationContext replicationContext) {
         super(corfuRuntime, session, replicationContext);
+        streamTagFollowed = LogReplicationUtils.SNAPSHOT_SYNC_QUEUE_TAG_SENDER_PREFIX + session.getSinkClusterId();
+
+        String snapshotSyncQueueFullyQualifiedName = TableRegistry.getFullyQualifiedTableName(CORFU_SYSTEM_NAMESPACE,
+                SNAPSHOT_SYNC_QUEUE_NAME_SENDER);
+        snapshotSyncQueueId = CorfuRuntime.getStreamID(snapshotSyncQueueFullyQualifiedName);
+
+        lastReadTimestamp = snapshotTimestamp;
+
+        String replicatedQueueName = TableRegistry.getFullyQualifiedTableName(CORFU_SYSTEM_NAMESPACE,
+                LogReplicationUtils.REPLICATED_QUEUE_NAME_PREFIX + session.getSourceClusterId());
+        replicatedQueueId = CorfuRuntime.getStreamID(replicatedQueueName);
+
+        try {
+            // Open the marker table so that its entries can be deserialized
+            CorfuStore corfuStore = new CorfuStore(replicationContext.getConfigManager().getRuntime());
+            corfuStore.openTable(CORFU_SYSTEM_NAMESPACE, SNAPSHOT_END_MARKER_TABLE_NAME,
+                Queue.RoutingTableSnapshotEndKeyMsg.class, Queue.RoutingTableSnapshotEndMarkerMsg.class, null,
+                TableOptions.fromProtoSchema(Queue.RoutingTableEntryMsg.class));
+        } catch(Exception e) {
+            log.error("Exception: ", e);
+        }
+
+        String endMarkerTableName = TableRegistry.getFullyQualifiedTableName(CORFU_SYSTEM_NAMESPACE,
+            SNAPSHOT_END_MARKER_TABLE_NAME);
+        endMarkerStreamId = CorfuRuntime.getStreamID(endMarkerTableName);
+
+        buildOpaqueStreamIterator();
+    }
+
+    // Create an opaque stream and iterator for the stream of interest, starting from lastReadTimestamp+1 to the global
+    // log tail
+    private void buildOpaqueStreamIterator() {
+        OpaqueStream opaqueStream =
+                new OpaqueStream(rt.getStreamsView().get(CorfuRuntime.getStreamID(streamTagFollowed)));
+
+        // Seek till the last read timestamp so that duplicate entries are eliminated.
+        opaqueStream.seek(lastReadTimestamp + 1);
+
+        long logTail = rt.getAddressSpaceView().getLogTail();
```

Comment:
Another way can be use the IStremView's next() api. That fetches the data from cache if present else fetches the new tail and refreshes the cache.
This way we'll not need to create a new iterator, the queryTask can poll on `next()`
But Opaque entry doesn't have a next() api right now, we'll need to add it.

Reply:
so the flow would be: create `opaqueStream -> seek -> next()`